network-api/networkapi/wagtailpages/pagemodels/index_page.py

```diff
@@ -0,0 +1,300 @@
+        context['index_title'] = titlecase(f'{category.name} {self.title}')
+        context['index_intro'] = category.intro
+
+        # and then the filtered content
+        context['terms'] = [category.name, ]
+        entries = [
+            entry
+            for
+            entry in entries.specific()
+            if
+            hasattr(entry, 'category')
+            and
+            category in entry.category.all()
+        ]
+
+        return entries
+
+    """
+    Sub routes
+    """
+
+    @route('^entries/')
+    def generate_entries_set_html(self, request, *args, **kwargs):
+        """
+        JSON endpoint for getting a set of (pre-rendered) entries
+        """
+
+        page = 1
+        if 'page' in request.GET:
+            page = int(request.GET['page'])
+
+        page_size = self.page_size
+        if 'page_size' in request.GET:
+            page_size = int(request.GET['page_size'])
+
+        start = page * page_size
+        end = start + page_size
+        entries = self.get_entries()
+        has_next = end < len(entries)
+
+        hide_classifiers = False
+        if hasattr(self, 'filtered'):
+            if self.filtered.get('type') == 'category':
+                hide_classifiers = True
+
+        html = loader.render_to_string(
+            'wagtailpages/fragments/entry_cards.html',
+            context={
+                'entries': entries[start:end],
+                'hide_classifiers': hide_classifiers
+            },
+            request=request
+        )
+
+        return JsonResponse({
+            'entries_html': html,
+            'has_next': has_next,
+        })
+
+    """
+    tag routes
+    """
+
+    # helper function for /tags/... subroutes
+    def extract_tag_information(self, tag):
+        terms = list(filter(None, re.split('/', tag)))
+        self.filtered = {
+            'type': 'tags',
+            'terms': terms
+        }
+
+    @route(r'^tags/(?P<tag>.+)/entries/')
+    def generate_tagged_entries_set_html(self, request, tag, *args, **kwargs):
+        """
+        JSON endpoint for getting a set of (pre-rendered) tagged entries
+        """
+        self.extract_tag_information(tag)
+        return self.generate_entries_set_html(request, *args, **kwargs)
+
+    @route(r'^tags/(?P<tag>.+)/')
+    def entries_by_tag(self, request, tag, *args, **kwargs):
+        """
+        If this page was called with `/tags/...` as suffix, extract
+        the tags to filter prior to rendering this page. Multiple
+        tags are specified as subpath: `/tags/tag1/tag2/...`
+        """
+        self.extract_tag_information(tag)
+        return IndexPage.serve(self, request, *args, **kwargs)
+
+    """
+    category routes
+    """
+
+    # helper function to resolve category slugs to actual objects
+    def get_category_object_for_slug(self, category_slug):
+        # FIXME: this should not need to be an app consult
+        BlogPageCategory = apps.get_model('wagtailpages', 'BlogPageCategory')
```

Comment:
This is the only line of "new" code, because we can't have an `from ..models import BlogPageCategory` at the top of this file, since this file gets imported _by_ `wagtailpages.models`, so that would lead to a circular import dependency. Instead we get that model "whenever this def gets called" by asking the Django app for it by name.